core/src/test/java/org/apache/calcite/plan/RelOptUtilTest.java

```diff
@@ -130,6 +130,43 @@
         Util.toLinux(RelOptUtil.dumpType(t2) + "\n"));
   }
 
+  @Test void testTypeDifference() {
+    final RelDataTypeFactory typeFactory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RelDataType t1 =
+        typeFactory.builder()
+            .add("f0", SqlTypeName.DECIMAL, 5, 2)
+            .add("f1", SqlTypeName.VARCHAR, 10)
+            .build();
+
+    RelDataType t2 =
+        typeFactory.builder()
+            .add("f0", SqlTypeName.DECIMAL, 5, 2)
+            .add("f1", SqlTypeName.VARCHAR, 5)
+            .build();
+
+    TestUtil.assertEqualsVerbose(
+        TestUtil.fold(
+            "Type mismatch: ",
+            "f1: VARCHAR(10) NOT NULL -> VARCHAR(5) NOT NULL",
+            ""),
+        Util.toLinux(RelOptUtil.getFullTypeDifferenceString(t1, t2) + "\n"));
+
+    t2 =
+        typeFactory.builder()
+            .add("f0", SqlTypeName.DECIMAL, 4, 2)
+            .add("f1", SqlTypeName.BIGINT)
+            .build();
+
+    TestUtil.assertEqualsVerbose(
+        TestUtil.fold(
+            "Type mismatch: ",
+            "f0: DECIMAL(5, 2) NOT NULL -> DECIMAL(4, 2) NOT NULL",
+            "f1: VARCHAR(10) NOT NULL -> BIGINT NOT NULL",
+            ""),
+        Util.toLinux(RelOptUtil.getFullTypeDifferenceString(t1, t2) + "\n"));
```

Comment:
This is nice, however, can you please add a planning case as well to see which message comes out of a planner?